src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java

```diff
@@ -1251,12 +1235,16 @@ private void handleConnectionFrame(Http2Frame frame)
             case AltSvcFrame.TYPE -> processAltSvcFrame(0, (AltSvcFrame) frame,
                     connection, connection.client());
 
-            default -> protocolError(ErrorFrame.PROTOCOL_ERROR);
+            default -> protocolError(ErrorFrame.PROTOCOL_ERROR, "unknown frame: " + frame);
         }
     }
 
-    boolean isOpen() {
-        return !isMarkedForShutdown() && connection.channel().isOpen();
+    /**
+     * Returns true if this connection hasn't been terminated and the underlying
+     * {@linkplain NetworkChannel#isOpen() channel is open}. false otherwise.
+     */
+    final boolean isOpen() {
+        return this.connTerminator.terminationCause.get() == null && connection.channel().isOpen();
```

Comment:
Can we ever observe a situation where channel is not open but termination cause is not set?

As far as I could tell, channel.isOpen only returns false after close() is called, and close() is only called from doTerminate after the termination cause is set. What am I missing?